### PR TITLE
Add -docker-link + -container-join-network flags (always connect networks/containers to arbitrary container ID/name)

### PR DESCRIPTION
**Pull request by CpuID**
OK here comes the potentially controversial, but also potentially very valuable feature/s, `-docker-link` + `-container-join-flags` :)

### Rationale

If you ever run Docker on a host, you may have a need to spawn some out-of-band process on the host OS, and connect to it from within the container. This is commonly done on the `172.17.0.1` for default bridge networks, or the default gateway for user defined bridge networks.

In a CI environment this can occur too, especially when you have Docker installed on the host, and you are running something like Jenkins (or others) on the same host. Jenkins may spawn some process which listens on some TCP/UDP port, and the container would connect back to the host.

In theory this is an anti-pattern (why not just run the out-of-band process in a sidecar container?), but there are valid use cases for it. For example, spawning an SSH tunnel from the host without `cp`-ing your SSH key into the host and connecting to it from the container, or using the [SSH agent TCP proxy pattern](https://github.com/aidanhs/docker-sshagent-socket) etc.

These features allow you to specify your CI container ID/name (detected via out-of-band logic, bash etc) and always join it to created containers via sockguard using Docker links and/or networks, depending on the architecture. You can then either use the `-docker-link` alias, or the CI container hostname to simulate talking to the "host" system that spawned the container, filling that void.

This eliminates the need for ugly hacks like having your CI container listen on some published port, and having your spawned containers go container -> host -> CI container, it will rely on native Docker networks to fulfil that traffic path instead.

Amazon ECS does not support [user defined networks yet](https://github.com/aws/amazon-ecs-agent/issues/437) (!), but other schedulers (eg. Kubernetes) I assume do (since `--link` has been deprecated for an extended period, but is still in widespread use). In that case, I supported both approaches for scheduler flexibility.

### `-docker-link`

Allows you to specify something like:

```
$ ./sockguard -docker-link festive_agnesi:testcicontainer -debug
Dialing directly
Adding a Docker --link to new containers: 'festive_agnesi:testcicontainer'
Listening on sockguard.sock (socket UID 1000 GID 1000 permissions 0600), upstream is /var/run/docker.sock
```

After that, any containers that are created via the guarded socket will have a `Links` entry of `festive_agnesi:testcicontainer`.

If a new Docker user defined network is spawned, this also connects the `festive_agnesi` to that new network (equivalent to `docker network connect newnetwork festive_agnesi`), and disconnects it on network delete API calls (to avoid "there are active endpoints" error). This is necessary for `docker-compose` support, since Compose puts all containers for a single "stack" in a new user defined bridge network by default, and for Compose external `links` support they must be connected to the user defined network:

https://docs.docker.com/compose/compose-file/compose-file-v2/#external_links

> Note: For version 2 file format, the externally-created containers must be connected to at least one of the same networks as the service which is linking to them.

### `-container-join-network`

Allows you to specify something like:

```
$ ./sockguard -container-join-network festive_agnesi -debug
Dialing directly
Container 'festive_agnesi' will always be connected to user defined bridged networks created via sockguard
Listening on sockguard.sock (socket UID 1000 GID 1000 permissions 0600), upstream is /var/run/docker.sock
```

After that, any networks that are created via sockguard will automatically have the `festive_agnesi` container connected, and disconnected prior to network delete respectively (to avoid "there are active endpoints" error).

### `-container-join-network-alias`

Only permitted if `container-join-network` is specified, will apply a selected hostname alias to the container joined to all networks.

Rationale: a container scheduler will spawn some programmatically generated container name and/or container ID SHA, and you may want a predictable name (eg. `buildkite-agent` or `jenkins-agent`) as an alias.

### Review Notes

@lox I know this PR is massive, and I was going to split it up more but in the end it would involve adding a lot of boilerplate into master that didn't get used, hence doing it atomically in one PR. I'm sure theres a bunch of places where some de-duplication could occur, but I realistically can't sink anymore time into this feature immediately (I'd like to tidy it up more in future though).

Test coverage: all passes, at 46.0%

### CODE REVIEW

**Comment by CpuID** [comment]
Hold off on a review straight away here @lox - I'm going to add an extra feature (the ability to set a network alias when using `-container-join-network`, similiar to the second `:` delimited arg for `-docker-link`).

**Comment by lox** [comment]
This is looking pretty good @CpuID. My main feedback will be around the docs and the top-level readme, and perhaps the wording of the parameters. Lemme know when you want me to review in depth, I'll review my review check in the meantime.

**Comment by CpuID** [comment]
> This is looking pretty good @CpuID. My main feedback will be around the docs and the top-level readme, and perhaps the wording of the parameters. Lemme know when you want me to review in depth, I'll review my review check in the meantime.

Thx, sounds good :) Yea it probably justifies a specific readme mention, I also have a bash snippet we used to detect the container we want to link to (our CI agent) in ECS, which is probably a useful example.

Yea I pondered on the CLI arg naming at least 3 times, so I'm definitely open to revision there. For now I left it alone.

I'll try get that last feature sorted in the next day or two.

**Comment by CpuID** [comment]
Note to self (incase I don't get to this today): the network join with alias feature looks ready, and tests pass. I need to give this a non-automated trial run to ensure this behaves as expected before handover for review.

**Comment by CpuID** [comment]
@lox this is ready for a more thorough review, unit tests all pass + gave it a dry run locally and it behaves as expected.

Happy to discuss changing the naming / wording of things, I'm also not 100% on it myself but it functions as it should :)

File: director.go [hunk 1]
```diff
@@ -115,12 +119,20 @@ func (r *rulesDirector) Direct(l socketproxy.Logger, req *http.Request, upstream
...
+		return r.handleNetworkDelete(l, req, upstream)
 	case match(`GET`, `^/networks/(.+)$`),
-		match(`DELETE`, `^/networks/(.+)$`),
 		match(`POST`, `^/networks/(.+)/(connect|disconnect)$`):
+		defer req.Body.Close()
+		connectBody, err := ioutil.ReadAll(req.Body)
+		if err != nil {
+			return errorHandler(err.Error(), http.StatusInternalServerError)
+		}
+		fmt.Printf("network connect body: %s\n", connectBody)
```
Comment (lox) [inline director.go]: Did you mean to leave this in here?

Comment (CpuID) [reply director.go]: nope :) thx for spotting, removed

**Comment by CpuID** [comment]
@lox *bump* :)

To confirm - we have this in live use, with success so far.

**Comment by lox** [comment]
Yup, let's roll with this for now. My hesitations were mostly around the precedent it sets. I'm still not sure it fits with "sockguard" as a concept, it's more like "socket proxy that adds constraints and makes changes to the docker environment transparently", but I think that is ok.

When I get a chance, I'll help out with some docs for this and perhaps look at if there is any refactoring we can do around moving some bulk out of director.go, but really appreciate the work you've put in here!

**Comment by CpuID** [comment]
> Yup, let's roll with this for now. My hesitations were mostly around the precedent it sets. I'm still not sure it fits with "sockguard" as a concept, it's more like "socket proxy that adds constraints and makes changes to the docker environment transparently", but I think that is ok.

Ya I agree in terms of "how much do we want to get in the way", definitely the less the better. In terms of "where this functionality fits", I think sockguard is the ideal place, alternatively it would need a second proxy implementation to do it independently. And I think since its default off it's a pretty reasonable feature.

> When I get a chance, I'll help out with some docs for this and perhaps look at if there is any refactoring we can do around moving some bulk out of director.go, but really appreciate the work you've put in here!

sounds good, yea director.go is probably to a point it could justify being broken out :)

thx for merging.

**Pull request merged**